Throwing ServerErrorException causes ClassNotFoundException

If, for instance, a JPA entity does not match the underlying database a `java.sql.SQLSyntaxErrorException` is thrown. This exception is handled by:

```
    private static RuntimeException handleInvocationTargetException(InvocationTargetException e) {
        Throwable exception = e.getTargetException();
        if (exception instanceof HttpStatusException || exception instanceof WebApplicationException) {
            return (RuntimeException) exception;
        }
        log.debug("Caught an unexpected exception (rethrowing as internal server error)", e);
        return new ServerErrorException("Unexpected exception caught", Response.Status.INTERNAL_SERVER_ERROR, e);
    }
```

If the application has a dependency to `org.glassfish.jersey.core:jersey-common`, this results in the following response:

```
{"errors":[{"status":"500","title":"javax.ws.rs.ServerErrorException","detail":"Unexpected exception caught"}]}
```

If not, constructing a `ServerErrorException` causes:

```
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: org.glassfish.jersey.internal.RuntimeDelegateImpl
	at javax.ws.rs.ext.RuntimeDelegate.findDelegate(RuntimeDelegate.java:152)
	at javax.ws.rs.ext.RuntimeDelegate.getInstance(RuntimeDelegate.java:120)
	at javax.ws.rs.core.Response$ResponseBuilder.newInstance(Response.java:848)
	at javax.ws.rs.core.Response.status(Response.java:590)
	at javax.ws.rs.core.Response.status(Response.java:601)
	at javax.ws.rs.ServerErrorException.<init>(ServerErrorException.java:158)
	at com.yahoo.elide.core.PersistentResource.handleInvocationTargetException(PersistentResource.java:1914)
	at com.yahoo.elide.core.PersistentResource.getValue(PersistentResource.java:1618)
	at com.yahoo.elide.datastores.hibernate5.HibernateTransaction.getRelation(HibernateTransaction.java:191)
	at com.yahoo.elide.core.PersistentResource.getRelationUnchecked(PersistentResource.java:1022)
	at com.yahoo.elide.core.PersistentResource.getRelation(PersistentResource.java:886)
	at com.yahoo.elide.core.PersistentResource.getRelationCheckedFiltered(PersistentResource.java:863)
	at com.yahoo.elide.core.PersistentResource.getRelationshipsWithRelationshipFunction(PersistentResource.java:1271)
	at com.yahoo.elide.core.PersistentResource.getRelationships(PersistentResource.java:1245)
	at com.yahoo.elide.core.PersistentResource.toResource(PersistentResource.java:1234)
	at com.yahoo.elide.core.PersistentResource.toResource(PersistentResource.java:1211)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.lambda$addResourcesForPath$3(IncludedProcessor.java:81)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.addResourcesForPath(IncludedProcessor.java:80)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.lambda$null$1(IncludedProcessor.java:65)
	at java.util.Arrays$ArrayList.forEach(Arrays.java:3880)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.lambda$addIncludedResources$2(IncludedProcessor.java:63)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.addIncludedResources(IncludedProcessor.java:60)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.lambda$execute$0(IncludedProcessor.java:50)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at com.yahoo.elide.jsonapi.document.processors.IncludedProcessor.execute(IncludedProcessor.java:49)
	at com.yahoo.elide.parsers.state.CollectionTerminalState.handleGet(CollectionTerminalState.java:72)
	at com.yahoo.elide.parsers.state.StateContext.handleGet(StateContext.java:110)
	at com.yahoo.elide.parsers.GetVisitor.visitQuery(GetVisitor.java:34)
	at com.yahoo.elide.parsers.GetVisitor.visitQuery(GetVisitor.java:21)
	at com.yahoo.elide.generated.parsers.CoreParser$QueryContext.accept(CoreParser.java:558)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at com.yahoo.elide.generated.parsers.CoreBaseVisitor.visitStart(CoreBaseVisitor.java:20)
	at com.yahoo.elide.parsers.BaseVisitor.visitStart(BaseVisitor.java:47)
	at com.yahoo.elide.parsers.BaseVisitor.visitStart(BaseVisitor.java:33)
	at com.yahoo.elide.generated.parsers.CoreParser$StartContext.accept(CoreParser.java:107)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at com.yahoo.elide.Elide.lambda$get$0(Elide.java:90)
	at com.yahoo.elide.Elide.handleRequest(Elide.java:202)
	at com.yahoo.elide.Elide.get(Elide.java:85)
```

And the original exception is lost.

--- a/.travis.yml
+++ b/.travis.yml
@@ -17,7 +17,7 @@
   branches:
     only:
       - master
-      - elide-3.0-wip
+      - elide-3.x
       - "/^[0-9]+\\.[0-9]+\\.[0-9]+/"
   # build steps
   before_install:

--- a/changelog.md
+++ b/changelog.md
@@ -1,4 +1,9 @@
 # Change Log
+## 3.2.1
+**Fixes**
+ * PersistentResource now throws an exception that Elide can handle internally, rather than delegating
+ error handling to the container.
+
 ## 3.2.0
 **Features**
  * Updated interface to beta standalone application. Plans to finalize this before Elide 4.0 release.

--- a/elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java
+++ b/elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java
@@ -47,10 +47,7 @@ import org.apache.commons.collections4.CollectionUtils;
 import org.apache.commons.lang3.StringUtils;
 
 import javax.persistence.GeneratedValue;
-import javax.ws.rs.ServerErrorException;
 import javax.ws.rs.WebApplicationException;
-import javax.ws.rs.core.Response;
-
 import java.io.Serializable;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.AccessibleObject;

--- a/elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java
+++ b/elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java
@@ -1911,6 +1911,6 @@ public class PersistentResource<T> implements com.yahoo.elide.security.Persisten
             return (RuntimeException) exception;
         }
         log.debug("Caught an unexpected exception (rethrowing as internal server error)", e);
-        return new ServerErrorException("Unexpected exception caught", Response.Status.INTERNAL_SERVER_ERROR, e);
+        return new InternalServerErrorException("Unexpected exception caught", e);
     }
 }

--- a/elide-core/src/main/java/com/yahoo/elide/core/exceptions/InternalServerErrorException.java
+++ b/elide-core/src/main/java/com/yahoo/elide/core/exceptions/InternalServerErrorException.java
@@ -18,7 +18,12 @@ public class InternalServerErrorException extends HttpStatusException {
     public InternalServerErrorException(String message) {
         super(HttpStatus.SC_INTERNAL_SERVER_ERROR, message);
     }
+
     public InternalServerErrorException(Throwable e) {
         super(HttpStatus.SC_INTERNAL_SERVER_ERROR, e.toString(), e, null);
+    }
+
+    public InternalServerErrorException(String message, Throwable e) {
+        super(HttpStatus.SC_INTERNAL_SERVER_ERROR, message, e, null);
     }
 }

